### ISSUE TITLE
[RFC / Feature] Standardized Multi-Agent RL Interface via PettingZoo (`AssumeParallelEnv`)

### ISSUE THREAD

**Issue by maurerle** [assume-framework/assume#849]
Just a quick draft as discussion material how the usage of PettingZoo could be done in Assume without creating too much overhead while keeping the benefits of not needing the custom implementations of various rl algorithms.

The key is that
1. assume can run without peetingZoo (just load torch weights and run normally)
2. for learning, PettingZoo has the clock and runs the `step` function, in which Assume does "its" step and returns with the next weights of the RL step.
3. The for loop in world is replaced in training, but can still be used in evaluation mode easily

This does not look like too many changes as the current architecture matches the usage in pettingzoo quite well (step, reset..)

I guess that this will allow to run the PettingZoo in parallel on the same machine, which would also provide a new way parallelization (N assume simulations in parallel).

Take the following with a grain of salt and as a rough architectural write down which might not hold in reality :)

## 1. Summary & Motivation
Currently, ASSUME implements its own Reinforcement Learning (RL) stack (algorithms like MATD3/MAPPO, replay buffers, training loops, neural network management, and complex async cache synchronization in `LearningRole`).

Maintaining custom RL algorithm implementations creates significant maintenance overhead, introduces subtle bugs (e.g., buffer transformation issues, tensor device handling, NaN gradients), and limits flexibility to adopt new SOTA algorithms.

**Goal:** Decouple RL algorithm execution from the ASSUME simulation by providing a standard **PettingZoo `ParallelEnv`** wrapper.
* **RL Frameworks (TorchRL, CleanRL, Ray RLlib, Tianshou):** Handle neural networks, training loops, replay buffers, exploration noise, and GPU optimization.
* **ASSUME:** Remains focused purely on energy market simulation, bid formulation, clearing mechanisms, and domain-specific reward calculations.

---

### A. High-Level Architecture

```mermaid
flowchart TD
    subgraph External["1. External RL Framework (TorchRL / CleanRL / RLlib)"]
        Policy["Policy Networks (Actor/Critic)"]
        Buffer["Replay Buffer / DataCollector"]
        Optimizer["Loss Computation & Optimizer"]
    end

    subgraph Bridge["2. Gym / PettingZoo Interface Layer"]
        Env["AssumeParallelEnv (pettingzoo.ParallelEnv)"]
    end

    subgraph ASSUME["3. ASSUME Core Engine (Discrete Event Simulation)"]
        World["World & Mango Container"]
        Market["Market Operators & Clearing"]
        RLUnits["RL Units (RLBiddingStrategy)"]
        HeuristicUnits["Heuristic / Merit-Order Units"]
    end

    Buffer -- "1. env.step(actions)" --> Env
    Env -- "2. set_action(action)" --> RLUnits
    Env -- "3. advance clock / World._step()" --> World
    World --> Market
    RLUnits --> Market
    HeuristicUnits --> Market
    Market -- "4. clearing & dispatch" --> RLUnits
    RLUnits -- "5. get_obs() & calc_reward()" --> Env
    Env -- "6. (obs, rewards, dones, info)" --> Buffer
```

---

### B. Synchronous Step Sequence

Rather than running ASSUME in a decoupled background thread, the **RL framework drives the simulation step-by-step**:

$$1 \text{ PettingZoo step} \iff \Delta t \text{ discrete time advancement in ASSUME's Mango Clock}$$

```mermaid
sequenceDiagram
    autonumber
    participant RL as RL Framework (TorchRL / CleanRL)
    participant Env as AssumeParallelEnv (Wrapper)
    participant Unit as RL Unit Strategy
    participant Market as ASSUME Simulation (World / Markets)

    RL->>Env: env.step(actions: dict[str, np.ndarray])
    Env->>Unit: set_action(action)
    Env->>Market: advance_clock_to_next_decision() / World._step()
    Note over Market: Mango Container ticks: Bids submitted & Market clears
    Market->>Unit: Market results & dispatch assigned
    Env->>Unit: get_observation()
    Unit-->>Env: obs_dict
    Env->>Unit: calculate_reward()
    Unit-->>Env: reward_dict
    Env-->>RL: (obs, rewards, terminations, truncations, infos)
    Note over RL: Policy update / Replay Buffer storage
```

---

### C. Training vs. Inference in Contrast

A critical design goal is that **inference (production scenario execution) must not require PettingZoo or external RL training loops**.

```mermaid
flowchart LR
    subgraph Training["Training Mode (External Loop)"]
        direction TB
        T1["RL Framework (TorchRL / CleanRL)"] -->|env.step| T2["AssumeParallelEnv"]
        T2 -->|Gradient Updates| T3["Trained Checkpoint (actor.pt / onnx)"]
    end

    subgraph Inference["Inference Mode (Native ASSUME)"]
        direction TB
        I1["config.yaml (PretrainedRLStrategy)"] -->|Loads Model| I2["world.run() / CLI"]
        I2 -->|Native Run in 1 Go| I3["SQL / CSV Results"]
    end

    T3 -.->|Frozen Weights| I1
```

| Aspect | Training Mode (RL Research) | Inference Mode (Production / Scenario Analysis) |
| :--- | :--- | :--- |
| **Orchestrator / Driver** | External RL Framework (`TorchRL`, `CleanRL`, `RLlib`) | Native ASSUME Simulation (`World.run()` / CLI `assume run`) |
| **Stepping Mechanism** | Discretely driven step-by-step via `env.step(actions)` | Continuous execution from $t_{\text{start}}$ to $t_{\text{end}}$ in one go |
| **PettingZoo Required?** | **Yes** (used as the environment wrapper) | **No** (pure ASSUME + lightweight PyTorch/ONNX inference) |
| **Exploration Noise** | Enabled (e.g. Gaussian / Action Noise for exploration) | **Disabled** (100 % deterministic model evaluation) |
| **Gradients & Buffer** | Active (Replay buffer, loss backprop, optimizer updates) | **None** (zero training overhead, frozen weights) |
| **Multi-Agent Setup** | Simultaneous action batching across all RL units | Each unit calls its own loaded model in `calculate_bids()` |
| **Heuristic Units** | Coexist seamlessly (Merit-Order / Naive strategies) | Coexist seamlessly (Merit-Order / Naive strategies) |

---

### 3.1. Streamlined RL Unit Strategy (`EnergyRLBiddingStrategy`)
Remove neural network, device, noise, and cache logic from strategies. Keep only domain logic and Gymnasium spaces:

```python
class EnergyRLBiddingStrategy(BaseStrategy):

    @property
    def observation_space(self) -> gymnasium.spaces.Box:
        return gymnasium.spaces.Box(
            low=-np.inf, high=np.inf, shape=(38,), dtype=np.float32
        )

    @property
    def action_space(self) -> gymnasium.spaces.Box:
        return gymnasium.spaces.Box(
            low=-1.0, high=1.0, shape=(2,), dtype=np.float32
        )

    def set_action(self, action: np.ndarray):
        """Called by AssumeParallelEnv before auction gate closure."""
        self.current_action = action

    def calculate_bids(
        self, unit, market_config, product_tuples
    ) -> list[dict]:
        """Translates self.current_action into P_min / P_max orderbook bids."""
        bid_prices = self.current_action * self.max_bid_price
        return [
            {"price": bid_prices[0], "volume": unit.min_power, ...},
            {
                "price": bid_prices[1],
                "volume": unit.max_power - unit.min_power,
                ...,
            },
        ]

    def get_observation(self, unit, start, end) -> np.ndarray:
        """Constructs observation vector from forecasts and unit state."""
        return self.create_observation(unit, start, end)

    def calculate_reward(self, unit) -> float:
        """Calculates scalar reward after clearing (profit - costs - regret)."""
        return profit - operational_costs - regret
```

---

### 3.2. PettingZoo Wrapper (`AssumeParallelEnv`)
Located in `assume/reinforcement_learning/env.py`:

```python
import gymnasium as gym
import numpy as np
from pettingzoo import ParallelEnv
from assume.world import World


class AssumeParallelEnv(ParallelEnv):
    metadata = {"name": "assume_v1"}

    def __init__(self, config_path: str):
        super().__init__()
        self.world = World.from_config(config_path)
        self.world.setup()

        self.possible_agents = [
            u_id
            for u_id, u in self.world.units.items()
            if hasattr(u.strategy, "set_action")
        ]
        self.agents = list(self.possible_agents)

        self.observation_spaces = {
            a: self.world.units[a].strategy.observation_space
            for a in self.possible_agents
        }
        self.action_spaces = {
            a: self.world.units[a].strategy.action_space
            for a in self.possible_agents
        }

    def reset(self, seed: int | None = None, options: dict | None = None):
        """Resets ASSUME to simulation start (t=0) and returns initial observations."""
        self.agents = list(self.possible_agents)
        self.world.reset_simulation(seed=seed)

        observations = {
            a: self.world.units[a].strategy.get_observation(
                self.world.units[a], self.world.clock.time, ...
            )
            for a in self.agents
        }
        infos = {a: {} for a in self.agents}
        return observations, infos

    def step(self, actions: dict[str, np.ndarray]):
        """Steps ASSUME by one decision interval."""
        # 1. Distribute actions to RL strategies
        for agent_id, action in actions.items():
            self.world.units[agent_id].strategy.set_action(action)

        # 2. Advance simulation until next decision/clearing point
        # (uses existing World._step(container) under the hood)
        self.world.advance_to_next_decision_point()

        # 3. Collect new observations & rewards
        observations = {
            a: self.world.units[a].strategy.get_observation(
                self.world.units[a], self.world.clock.time, ...
            )
            for a in self.agents
        }
        rewards = {
            a: self.world.units[a].strategy.calculate_reward(
                self.world.units[a]
            )
            for a in self.agents
        }

        # 4. Check termination
        is_done = self.world.clock.time >= self.world.end_timestamp
        terminations = {a: is_done for a in self.agents}
        truncations = {a: False for a in self.agents}
        infos = {
            a: {"profit": self.world.units[a].last_profit} for a in self.agents
        }

        if is_done:
            self.agents = []

        return observations, rewards, terminations, truncations, infos
```

---

### 3.3. Standalone Pretrained Strategy for Inference (`PretrainedRLStrategy`)
For running trained policies natively in standard ASSUME simulations without PettingZoo:

```python
import torch


class PretrainedRLStrategy(EnergyRLBiddingStrategy):
    """Loads and executes a trained policy model deterministically inside ASSUME."""

    def __init__(self, model_path: str, **kwargs):
        super().__init__(**kwargs)
        # Load TorchScript, ONNX runtime, or PyTorch weights
        self.model = torch.jit.load(model_path)
        self.model.eval()

    def calculate_bids(
        self, unit, market_config, product_tuples
    ) -> list[dict]:
        start = product_tuples[0][0]
        end = product_tuples[0][1]

        # 1. Build observation
        obs = self.get_observation(unit, start, end)

        # 2. Deterministic inference (no gradients, no noise)
        with torch.no_grad():
            obs_tensor = torch.as_tensor(obs, dtype=torch.float32)
            action = self.model(obs_tensor).cpu().numpy()

        # 3. Formulate standard bids using the action
        self.set_action(action)
        return super().calculate_bids(unit, market_config, product_tuples)
```

---

### A. Training Loop (with TorchRL)
```python
from torchrl.collectors import SyncDataCollector
from torchrl.envs.libs.pettingzoo import PettingZooWrapper
from assume.reinforcement_learning.env import AssumeParallelEnv

# 1. Wrap ASSUME environment
env = PettingZooWrapper(
    AssumeParallelEnv("scenarios/day_ahead_multi_storage.yaml")
)

# 2. TorchRL collector automatically steps the environment & manages batches
collector = SyncDataCollector(
    env, policy, frames_per_batch=24, total_frames=24_000
)

# 3. Policy gradient updates
for batch in collector:
    loss = loss_module(batch)
    optimizer.step()
```

### B. Inference Option 1: Standalone Native CLI (`assume run`)
Users running production scenarios do **not** need PettingZoo or an RL loop:
```yaml
# config.yaml
units:
  battery_1:
    unit_type: "storage"
    strategy: "PretrainedRLStrategy"
    bidding_params:
      model_path: "models/trained_battery_actor.pt"
      foresight: 12
```
Command:
```bash
assume run scenarios/day_ahead_multi_storage.yaml
```

### C. Inference Option 2: Programmatic Evaluation Loop
For researchers running batch evaluations or sensitivity sweeps in Python:
```python
import torch
from assume.reinforcement_learning.env import AssumeParallelEnv

policy = torch.load("models/trained_battery_actor.pt")
policy.eval()

env = AssumeParallelEnv("scenarios/day_ahead_multi_storage.yaml")
obs, info = env.reset()

with torch.no_grad():
    while env.agents:
        actions = policy(obs)  # Deterministic action selection
        obs, rewards, terminations, truncations, infos = env.step(actions)
```

---

## 5. Implementation Tasks

- [ ] **Step 1: Refactor RL Strategies**
  - Extract NN/PyTorch/Buffer dependencies from `TorchLearningStrategy`.
  - Expose `observation_space`, `action_space`, `set_action()`, `get_observation()`, and `calculate_reward()`.
- [ ] **Step 2: Implement `AssumeParallelEnv`**
  - Implement `pettingzoo.ParallelEnv` wrapper calling `World._step()`.
  - Implement clean `reset_simulation()` method on `World`.
- [ ] **Step 3: Dependencies & Packaging**
  - Add `gymnasium` and `pettingzoo` to optional dependencies in `pyproject.toml` under `[project.optional-dependencies] rl = ["pettingzoo>=1.24.0", "gymnasium>=1.0.0"]`.
- [ ] **Step 4: Deprecate & Clean Up In-House RL Code**
  - Deprecate `assume/reinforcement_learning/algorithms/` (`matd3.py`, `mappo.py`).
  - Deprecate `assume/reinforcement_learning/buffer.py` and `LearningRole` internal cache synchronization.
- [ ] **Step 5: Integration Tests & Documentation**
  - Add pytest tests for `AssumeParallelEnv` reset, step consistency, and determinism.
  - Add tutorial/example script showing how to train with TorchRL / CleanRL.h

Related Links:

- https://benchmarl.readthedocs.io/en/latest/
- https://pettingzoo.farama.org/tutorials/agilerl/MATD3/
